posts/frangeris/03.secure-values-by-private-keys-in-aws.md

```diff
@@ -0,0 +1,235 @@
+    # we will need this value later
+    MyKeyId:
+      Value: !Ref MyPrivateKey
+      Description: The ID of the KMS key used to encrypt
+```
+
+In this stage you should be able to have the lambda function, a secret key which only the root user and the lambda ARN can access and use it. Now, how about creating our encrypted values?
+
+For simplicity I personally prefer [AWS CLI](https://aws.amazon.com/cli/) to handle the process of encrypting and saving instead of creating another lambda script that will need to use [AWS SDK](https://aws.amazon.com/sdk-for-javascript/), ending in the same result 🤷🏽
+
+So first, we need to encrypt our secret using the previous key already created, so that even if someone is able to access the value, they will need our key to be able to see the content,
+
+Some of the values used below are not exposed in the serverless deploy output, so to get those values is via the _stack output_ previously defined as part of `serverless.yml`:
+
+```bash
+npx sls info --verbose
+```
+
+> --plaintext (data to be encrypted) is required to be base64 text or a blob from a plain text file
+
+```bash
+aws kms encrypt \
+    --region <same region used by serverless> \
+    --key-id <key id from output (MyPrivateKeyId)> \
+    --plaintext "<base64 of value> or <fileb://path>" \
+    --output text \
+    --query CiphertextBlob
+```
+
+🎉 You should be able to get a base64 output from the command above with the encrypted value, now what next?
+
+### Store the values
+
+We need to save those values inside a secure service, either AWS System Manager (using Parameter Store) or AWS Secret Manager creating directly a key/value. Why not try both?
+
+#### For Parameter Store
+
+For reusability use the previous encrypted output directly with the same command, getting as result:
+
+```bash
+aws ssm put-parameter \
+  --name "my-parameter-name" \
+  --type "SecureString" \
+  --value "$(aws kms encrypt --region us-east-1 --key-id <key id> --plaintext "<value>" --output text --query CiphertextBlob | base64 --decode)" \
+  --region us-east-1
+```
+
+If everything works, you should receive a json payload like this, which means it works, and that the value was created successfully 👏🏽
+
+> Access via AWS Systems Manager -> Parameter Store to see it.
+
+```json
+{
+  "Version": 1,
+  "Tier": "Standard"
+}
+```
+
+#### For Secrets Manager
+
+Secret manager use the same approach as Parameter store, but with a different command, let's see:
+
+```bash
+aws secretsmanager create-secret --name "my-secret" --secret-string "my-secret-value"
+```
+
+Running this will creates a new secret in AWS Secrets Manager with the name `my-secret` and a secret value of `my-secret-value`, if the command is successful, it will return a JSON object containing metadata about the newly created secret, including its [Amazon Resource Name (ARN)](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html), let's encrypt and send it to `secretsmanager` as follow:
```

Comment:
```suggestion
Running this will create a new secret in AWS Secrets Manager with the name `my-secret` and a secret value of `my-secret-value`. If the command is successful, it will return a JSON object containing metadata about the newly created secret, including its [Amazon Resource Name (ARN)](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html).
Let's encrypt and send it to `secretsmanager` as follow:
```

Reply:
You got a tricky grammar here, thank you!